kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/LogSegment.java

```diff
@@ -176,6 +193,69 @@ public LogSegment(
                         Comparator.comparing((FileStatus a) -> new Path(a.getPath()).getName())
                             .reversed())
                     .collect(Collectors.toList()));
+
+    this.compactionsReversed =
+        new Lazy<>(
+            () ->
+                compactions.stream()
+                    .sorted(
+                        Comparator.comparing((FileStatus a) -> new Path(a.getPath()).getName())
+                            .reversed())
+                    .collect(Collectors.toList()));
+
+    this.allFilesWithCompactionsReversed =
```

Comment:
A few thoughts

1. Let's move this logic into a new private helper method `loadAllFilesWithCompactionsReversed`.

2. I think we should split this PR into two.
  a. One for gathering the list of compaction files (from the LIST) and
  b. Another for actually selecting the right compaction files for LogReplay. For example, this method alone (part b) could have a ton of unit tests for the various different edge cases